consensus/istanbul/backend/randao.go

```diff
@@ -0,0 +1,149 @@
+package backend
+
+import (
+	"bytes"
+	"math/big"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/klaytn/klaytn/accounts/abi/bind/backends"
+	"github.com/klaytn/klaytn/blockchain/system"
+	"github.com/klaytn/klaytn/blockchain/types"
+	"github.com/klaytn/klaytn/common"
+	"github.com/klaytn/klaytn/common/hexutil"
+	"github.com/klaytn/klaytn/consensus"
+	"github.com/klaytn/klaytn/crypto"
+	"github.com/klaytn/klaytn/crypto/bls"
+)
+
+// For testing without KIP-113 contract setup
+type BlsPubkeyProvider interface {
+	GetBlsPubkey(chain consensus.ChainReader, proposer common.Address, num *big.Int) (bls.PublicKey, error)
+	ResetBlsCache()
+}
+
+type ChainBlsPubkeyProvider struct {
+	cache *lru.ARCCache // Cached BlsPublicKeyInfos
+}
+
+func newChainBlsPubkeyProvider() *ChainBlsPubkeyProvider {
+	cache, _ := lru.NewARC(128)
+	return &ChainBlsPubkeyProvider{
+		cache: cache,
+	}
+}
+
+// The default implementation for BlsPubkeyFunc.
+// Queries KIP-113 contract and verifies the PoP.
+func (p *ChainBlsPubkeyProvider) GetBlsPubkey(chain consensus.ChainReader, proposer common.Address, num *big.Int) (bls.PublicKey, error) {
+	infos, err := p.getAllCached(chain, num)
+	if err != nil {
+		return nil, err
+	}
+
+	info, ok := infos[proposer]
+	if !ok {
+		return nil, errNoBlsPub
+	}
+	return bls.PublicKeyFromBytes(info.PublicKey)
+}
+
+func (p *ChainBlsPubkeyProvider) getAllCached(chain consensus.ChainReader, num *big.Int) (system.BlsPublicKeyInfos, error) {
+	if item, ok := p.cache.Get(num.Uint64()); ok {
+		logger.Trace("BlsPublicKeyInfos cache hit", "number", num.Uint64())
+		return item.(system.BlsPublicKeyInfos), nil
+	}
+
+	backend := backends.NewBlockchainContractBackend(chain, nil, nil)
```

Comment:
I don't think that's feasible because `chain` is passed down all the way from the `VerifyHeader`. In theory the `chain` can be different every time. Therefore it's unsafe to reuse one ContractBackend nor RegistryCaller.